File: mobile/lib/core/models/service_log.dart
class ServiceItem {
  final String name;
  final double cost;

  const ServiceItem({
    required this.name,
    required this.cost,
  });

  factory ServiceItem.fromJson(Map<String, dynamic> json) {
    return ServiceItem(
      name: json['name'] as String? ?? '',
      cost: (json['cost'] as num?)?.toDouble() ?? 0.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'cost': cost,
    };
  }

  ServiceItem copyWith({String? name, double? cost}) {
    return ServiceItem(
      name: name ?? this.name,
      cost: cost ?? this.cost,
    );
  }
}

class ServiceLog {
  final String id;
  final String bikeId;
  final String loggedAt;
  final double odometerReading;
  final List<ServiceItem> serviceItems;
  final double cost;
  final String? workshopName;
  final double? nextServiceKm;
  final String? nextServiceDate;
  final String? notes;
  final String createdAt;

  const ServiceLog({
    required this.id,
    required this.bikeId,
    required this.loggedAt,
    required this.odometerReading,
    required this.serviceItems,
    required this.cost,
    this.workshopName,
    this.nextServiceKm,
    this.nextServiceDate,
    this.notes,
    required this.createdAt,
  });

  factory ServiceLog.fromJson(Map<String, dynamic> json) {
    final itemsJson = json['service_items'] as List<dynamic>? ?? [];
    return ServiceLog(
      id: json['id']?.toString() ?? '',
      bikeId: json['bike_id']?.toString() ?? '',
      loggedAt: json['logged_at'] as String? ?? '',
      odometerReading: (json['odometer_reading'] as num?)?.toDouble() ?? 0.0,
      serviceItems: itemsJson
          .map((e) => ServiceItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      cost: (json['cost'] as num?)?.toDouble() ?? 0.0,
      workshopName: json['workshop_name'] as String?,
      nextServiceKm: (json['next_service_km'] as num?)?.toDouble(),
      nextServiceDate: json['next_service_date'] as String?,
      notes: json['notes'] as String?,
      createdAt: json['created_at'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bike_id': bikeId,
      'logged_at': loggedAt,
      'odometer_reading': odometerReading,
      'service_items': serviceItems.map((e) => e.toJson()).toList(),
      'cost': cost,
      if (workshopName != null) 'workshop_name': workshopName,
      if (nextServiceKm != null) 'next_service_km': nextServiceKm,
      if (nextServiceDate != null) 'next_service_date': nextServiceDate,
      if (notes != null) 'notes': notes,
      'created_at': createdAt,
    };
  }

  Map<String, dynamic> toCreateJson() {
    return {
      'bike_id': bikeId,
      'logged_at': loggedAt,
      'odometer_reading': odometerReading,
      'service_items': serviceItems.map((e) => e.toJson()).toList(),
      'cost': cost,
      if (workshopName != null) 'workshop_name': workshopName,
      if (nextServiceKm != null) 'next_service_km': nextServiceKm,
      if (nextServiceDate != null) 'next_service_date': nextServiceDate,
      if (notes != null) 'notes': notes,
    };
  }
}

File: backend/app/models/service_log.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Optional

from sqlalchemy import Float, Date, DateTime, ForeignKey, String
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.base import UUIDMixin, TimestampMixin


class ServiceLog(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "service_logs"

    bike_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("bikes.id", ondelete="CASCADE"), nullable=False, index=True
    )
    logged_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    odometer_reading: Mapped[Optional[float]] = mapped_column(Float)
    # [{name: str, cost: float}, ...] — individual line items
    service_items: Mapped[list] = mapped_column(JSONB, nullable=False)
    cost: Mapped[float] = mapped_column(Float, nullable=False)  # sum of service_items costs
    workshop_name: Mapped[Optional[str]] = mapped_column(String(200))
    next_service_km: Mapped[Optional[float]] = mapped_column(Float)
    next_service_date: Mapped[Optional[date]] = mapped_column(Date)
    notes: Mapped[Optional[str]] = mapped_column(String(1000))

    bike: Mapped["Bike"] = relationship("Bike", back_populates="service_logs")
